### ISSUE TITLE
optimize runtime varying integer divisions

### ISSUE THREAD

**Issue by MkazemAkhgary** [ispc/ispc#3001]
integer division can be optimized to a multiply and shift. this is done when divisor is known at compile time. but this technique is also beneficial at runtime in SIMD world. here are some cases where it can improve performance:

### 1. divide varying integer by a uniform integer
varying integer division with an N-wide gang can be reduced to just one division and some multiples and shifts instead of doing N individual divisions.

### 2. Loop invariant varying integer divisions
when doing varying integer divisions inside a loop and the divisor is loop invariant, magic quotient can be computed outside the loop to make integer divisions much faster inside the loop. this should work for both uniform and varying divisors that are loop invariant.

### 3. use FP units to divide varying integers.
I'm not sure if this works for all cases, but I think for example 16 bit integer division can be done via 32 bit floating point division, and 32 bit integer division can be done via 64 bit floating point division and this should still be faster than doing individual integer divisions especially for large gang sizes.

**Comment by pbrubaker** [comment]
>use FP units to divide varying integers.

In my opinion, I don't think the compiler should be making this decision transparently.  I think this should be explicitly implemented by the programmer.

That aside, there is a cost for converting between integer and floating point.

**Comment by zephyr111** [comment]
> varying integer division with an N-wide gang can be reduced to just one division and some multiples and shifts instead of doing N individual divisions.

This is a good idea indeed, but probably mainly for wide SIMD instruction sets (and if the division is not particularly optimized on the target platform) or/and also short integers. 

For example, on one hand, with Neon and 64-bit integers (i.e. `programCount` set to 2), 2 division instructions are needed per `varying/uniform` division so the pre-computing overhead plus the actual add/shifts instructions is certainly significantly higher than just computing the actual division. Put it shortly, in this case, it does not worth it. On the other-hand, for 32-bit divisions on AVX-512 (i.e. `programCount` set to 16), the overhead can surely be mitigated, especially with double-pumping. This assumes the final instructions replacing the divisions can fully benefit from the SIMD instruction set used (some `uniform` instructions do not have any efficient `varying` alternative regarding the target SIMD instruction set chosen).

On top of that, some modern CPUs, like the Apple's M1, integer divisions are actually pretty fast. Signed/unsigned division only take 2 cycles of throughput! Thus the benefit of using this pre-computation is certainly tiny (if any) on a single `varying/uniform` division.

Regarding the overhead, libdivide states the precomputation takes about the time of 1-3 divisions:
> Pre-computing the proper magic number and shift is on the order of one to three hardware divides, for non-powers of 2. Powers of 2 (and their negatives) are much faster. 

> [...] magic quotient can be computed outside the loop to make integer divisions much faster inside the loop

Yes, but the thing is the compiler (and sometimes the also the user) do not know how many iterations of the loop are going to be executed. For few iterations, it is certainly slower to do the pre-computation. ISPC can generate 2 versions (one with the pre-computation and one without) but it makes the code bigger and more complex and we should also set a magic threshold.

IMHO, this use-case is pretty uncommon compared to `varying/uniform` divisions. Indeed, I expect the divisors to often change in the loop if they are stored in a `varying`.

> use FP units to divide varying integers.

I agree with pbrubaker : I do not expect the compiler to do that automatically. That being said, ISPC could provide an optimization flag for that (similar to fast-math).

For quite small 32-bit integers (<24 bit), 32-floats can be used instead of 64-bit one and this is not something so easy for the compiler to do since most of the time it does know how many bits are actually used (unless users explicitly specify this which is rather uncommon). Thus, I think users can make better decision than the compiler with quite little effort in this case.

-----

To sum up
---

Overall, regarding the pre-computation for divisions, IMHO it is not the kind optimizations that can be blindly applied in all cases and it is hard for the compiler to make efficient decisions. I wonder if we could **wrap this in software functions** rather than compilers optimizations. The former is significantly more flexible and much easier to implement IMHO. Moreover, I expect users to make better decisions (with little effort) than what ISPC can do automatically here.

We should also consider the impact of adding a dependency like libdivide (on the size of binaries, distribution, licensing).

Regarding the FP-related division optimization, I think an *optional optimization flag* (disabled by default) is certainly the best solution assuming it really worth it. I must fit well with the libdivide-like optimization (also using integer divisions).

**Comment by nurmukhametov** [comment]
> > varying integer division with an N-wide gang can be reduced to just one division and some multiples and shifts instead of doing N individual divisions.
> 
> This is a good idea indeed, but probably mainly for wide SIMD instruction sets (and if the division is not particularly optimized on the target platform) or/and also short integers.
> 
> For example, on one hand, with Neon and 64-bit integers (i.e. `programCount` set to 2), 2 division instructions are needed per `varying/uniform` division so the pre-computing overhead plus the actual add/shifts instructions is certainly significantly higher than just computing the actual division. Put it shortly, in this case, it does not worth it. On the other-hand, for 32-bit divisions on AVX-512 (i.e. `programCount` set to 16), the overhead can surely be mitigated, especially with double-pumping. This assumes the final instructions replacing the divisions can fully benefit from the SIMD instruction set used (some `uniform` instructions do not have any efficient `varying` alternative regarding the target SIMD instruction set chosen).

> ... I wonder if we could **wrap this in software functions** rather than compilers optimizations. The former is significantly more flexible and much easier to implement IMHO. ...

Actually since #3322, varying divisions have been effectively implemented in ISPC (see [here](https://github.com/ispc/ispc/blob/841579e06a70da010888d214287a8b4b98bc171d/builtins/generic.ispc#L283)). We can quite easily make them target and width specific (using ifdefs with TARGET_WIDTH and ISPC_TARGET_* if needed).

**Comment by zephyr111** [comment]
> Actually since https://github.com/ispc/ispc/pull/3322, varying divisions have been effectively implemented in ISPC (see [here](https://github.com/ispc/ispc/blob/841579e06a70da010888d214287a8b4b98bc171d/builtins/generic.ispc#L283)).

Indeed. This is great if we wish to tweak de behavior of the division (so if we add an option in ISPC to automatically implement the behavior of libdivide).

-----

libdivide benchmark and implementations
---

I performed some tests/benchmarks on my i5-9600KF CPU (Intel Skylake with 2xAVX2 units):

```
                   === libdivide u16 benchmark ===

     #     system     scalar     scl_bf     vector     vec_bf      gener   algo
     1   1737.595    303.268     -1.000     51.498     -1.000    867.844    0
     2   1737.595    303.268    762.939     51.498     49.591    867.844    0
     3   1737.595    511.169    762.939     51.498     49.591    867.844    1
     4   1737.595    303.268    762.939     51.498     49.591    867.844    0
     5   1737.595    511.169    762.939     51.498     49.591    867.844    1
     6   1737.595    511.169    764.847     49.591     49.591    867.844    1
     7   1737.595    762.939    762.939     66.757     49.591    869.751    2
     8   1737.595    303.268    762.939     51.498     49.591    869.751    0
     9   1737.595    511.169    764.847     49.591     49.591    867.844    1
    10   1737.595    511.169    764.847     51.498     49.591    867.844    1

                   === libdivide s16 benchmark ===

     #     system     scalar     scl_bf     vector     vec_bf      gener   algo
     1   1737.595    690.460     -1.000     66.757     -1.000   1085.281    0
     2   1737.595    690.460    932.693     66.757     62.943   1085.281    0
     3   1737.595    864.029    944.138     72.479     62.943   1085.281    2
     4   1737.595    692.368    946.045     62.943     62.943   1085.281    0
     5   1737.595    621.796    930.786     55.313     62.943   1085.281    1
     6   1737.595    864.029    930.786     72.479     62.943   1085.281    2
     7   1737.595    623.703    928.879     57.220     62.943   1085.281    1
     8   1737.595    661.850    946.045     68.665     62.943   1085.281    0
     9   1737.595    864.029    928.879     74.387     62.943   1085.281    2
    10   1737.595    621.796    930.786     57.220     62.943   1085.281    1

                   === libdivide u32 benchmark ===

     #     system     scalar     scl_bf     vector     vec_bf      gener   algo
     1   1738.015    308.388     -1.000     81.856     -1.000    868.056    0
     2   1738.015    308.388    681.500     78.049    137.061    868.056    0
     3   1738.015    512.077    681.500    131.351    137.061    868.056    1
     4   1738.015    308.388    681.500     79.952    137.061    868.056    0
     5   1738.015    512.077    681.500    129.447    137.061    868.056    1
     6   1738.015    512.077    681.500    129.447    137.061    868.056    1
     7   1738.015    681.500    681.500    152.290    137.061    868.056    2
     8   1738.015    308.388    681.500     79.952    137.061    868.056    0
     9   1738.015    512.077    681.500    127.543    137.061    868.056    1
    10   1738.015    512.077    681.500    127.543    137.061    868.056    1

                   === libdivide s32 benchmark ===

     #     system     scalar     scl_bf     vector     vec_bf      gener   algo
     1   1738.433    630.776     -1.000    119.695     -1.000   1086.758    0
     2   1738.433    630.776    877.766    121.595    169.094   1086.758    0
     3   1738.433    788.470    877.766    169.094    169.094   1086.758    2
     4   1738.433    630.776    877.766    115.896    169.094   1086.758    0
     5   1738.433    623.176    877.766    146.294    169.094   1086.758    1
     6   1738.433    788.470    877.766    172.893    169.094   1086.758    2
     7   1738.433    788.470    877.766    172.893    169.094   1086.758    2
     8   1738.433    630.776    877.766    117.795    169.094   1086.758    0
     9   1738.433    623.176    877.766    144.394    169.094   1086.758    1
    10   1738.433    623.176    877.766    148.194    169.094   1086.758    1

                   === libdivide u64 benchmark ===

     #     system     scalar     scl_bf     vector     vec_bf      gener   algo
     1   5729.928    407.377     -1.000    152.290     -1.000    908.032    0
     2   5484.360    390.244    723.380    140.869    414.991    868.056    0
     3   5478.649    521.595    721.476    367.401    411.184    868.056    1
     4   5482.456    388.341    721.476    138.965    413.088    868.056    0
     5   5480.553    521.595    719.572    365.497    411.184    868.056    1
     6   5480.553    521.595    721.476    367.401    411.184    868.056    1
     7   5480.553    719.572    721.476    428.317    411.184    868.056    2
     8   5478.649    390.244    719.572    138.965    411.184    868.056    0
     9   5480.553    521.595    719.572    365.497    409.281    868.056    1
    10   5480.553    521.595    721.476    367.401    411.184    868.056    1

                   === libdivide s64 benchmark ===

     #     system     scalar     scl_bf     vector     vec_bf      gener   algo
     1   6042.124    633.909     -1.000    285.545     -1.000   1086.973    0
     2   6042.124    633.909    887.092    285.545    622.487   1086.973    0
     3   6042.124    788.103    888.995    601.547    622.487   1086.973    2
     4   6042.124    633.909    887.092    283.641    624.391   1086.973    0
     5   6042.124    618.680    887.092    548.246    624.391   1086.973    1
     6   6042.124    786.199    887.092    603.451    626.294   1086.973    2
     7   6049.738    635.813    900.417    544.438    637.716   1086.973    1
     8   6042.124    633.909    887.092    281.737    624.391   1085.069    0
     9   6042.124    786.199    888.995    601.547    622.487   1086.973    2
    10   6040.220    618.680    888.995    550.149    624.391   1085.069    1
```

The first column is the divisor used/pre-computed. The second one is the hardware divide time (i.e. `div`/`idiv` instruction). Others are the implementation (and `gener` is the time to do the pre-computation). `*_bf` columns are for branch-free implementations. Apparently, timings are per item and in ns (with many repetitions). Only the relative time matters, not the absolute one. See the [libdivide github](https://github.com/ridiculousfish/libdivide) for more information about their benchmark. 

I removed negative numbers from the benchmark since the performance is independent from the sign in practice. What matters for performance is the type of the integers and the algorithm used. There are apparently 3 algorithms used based on the actual divisor. For example, when the divisor is a power of two, then there is only a single shift needed in some implementations for unsigned divisions resulting in a much faster execution. Besides, please note that 8-bit divisions are not supported in libdivide yet.

Since the main purpose of writing ISPC code is to benefit from the performance of SIMD instruction and not really scalar one, let us focus on the vectorized implementation, and more specifically the branch-free one (i.e. `vec_bf`). Indeed, the branch-full one cannot be efficient when the divisor is a varying one. In fact, this is not actually implemented in libdivide (and will certainly never be).

Overall, the `vec_bf` implementation is significantly faster than the scalar ones for 16-bit and 32-bit integers, especially compared to the naive `div`/`idiv` instructions. For 64-bit types, it is not really faster than the scalar libdivide implementation using conditional branches (despite the use of 4 x 64-bit SIMD lanes). Note the branch-full SIMD version is significantly faster than the branch-free ones for power-of-two numbers since the branch-full version mainly just perform a basic shift while the branch-free need significantly more operations. I do not think this use-case is the most important though since developers can write optimized code using shifts themselves when they know that the divisor has a good chance to be a power of two or also when the divisor is known to be tiny. Moreover, it mostly matters for 64-bit integers while ISPC users used to manipulate 32-bit variables more frequently. By the way, please note that the branch-free version does not support divisions by 1.

Note that Skylake is known to have a pretty slow 64-bit divisions. Nowadays CPUs can compute it significantly faster (AFAIK about 3-4 times faster for IceLake, AlderLake P-cores, Zen4 and Zen4). On some CPU (e.g. Apple's M1), the scalar implementation of libdivide it only barely faster, especially the branch-free one (or the branch-less ones in pathological cases). Also please note that on Zen5 and Intel server-side CPUs, AVX-512 should make the `vec` and `vec_bf` versions even faster (up to twice more for 16-bit and 32-bit data types and possibly a bit more with 64-bit ones).

-----

libdivide code analysis
---

The code of libdivide is pretty big for what it does. A significant part is vectorized implementations (e.g. SSE2, AVX2, AVX-512, Neon) and the code used to pre-compute the divisions. A source of code bloat is the implementation of a function for several different types (i.e. signed/unsigned 16-bit/32-bit/64-bit integers) with some changes in the code for each variants. On top of that there are also multiple algorithms implemented.

Implementation decisions
---

Adding libdivide as a binary dependency is not great since LLVM cannot inline the compiled functions. This is critical for performance since a significant part of the function often compute constants and such instructions could be moved outside a loop if the code was inlined. Moreover, I wonder how this impact the packaging of the ISPC binaries. Additionally, some platforms/targets supported by ISPC will certainly not be supported by libdivide so a fallback is needed (e.g. double pumping targets needs to be considered). Last but not least, libdivide does not efficiently support the case where the divisor would be a varying integer which is a significant constraint since some quick benchmarks shows that it can be implemented so to be significantly faster than the naive scalar implementations (at least for 32-bit integers).

I think an alternative solution is to generate the LLVM IR code and not directly binaries so for LLVM to be able to inline functions. However, this does not address the last issues.

Yet another solution is to re-implement libdivide from scratch in ISPC. This solve most of the issue. However, this is not easy to do, it will certainly be a bit less efficient than the libdivide code using manually intrinsics, and it may results in a pretty big code. Regarding the later point, I wonder if it would impact build timing if we put that in the standard library (and so if it is a good idea to put this in it).

-----

Current implementation/prototype issues
---

I tried to write a naive ISPC prototype (based on the last solution), but ISPC templates are currently not powerful enough so to express the code of libdivide in a generic way (and macros are not great too). One main issue is that there are many variants to generates: 
 - `uniform`/`varying` numerator
 - `uniform`/`varying` denominator
 - `signed`/`unsigned` divisions
 - 16-bit, 32-bit and 64-bit integers

This means 24 versions for the function doing the actual division (i.e. "apply" function). There is also the function to pre-compute the division (i.e. "generate" function) and some utility functions. In the "apply" functions, the code for different integer size is mostly (but not totally) identical, the code of signed integers is pretty different from the unsigned ones and the code is nearly the same regarding the variability of the input variable.

Here is one practical example:

```ispc
// T is an integral type and PrecomputedDividerData is a custom struct.
template <typename T, typename PrecomputedDividerData>
ReturnType /* returns `varying T` if T is `varying` OR `PrecomputedDividerData` is `varying`  */ compute(T numerator, struct PrecomputedDividerData divider)
{
    TemporaryVarType1 var1 = ...; // The variability is dependent of the one of both T and PrecomputedDividerData
    signed T var2 = ...; // The type is an unsigned version of T which is supposed to be signed but `signed T` is not supported
    // ...
}
```

Actually, a similar issue is found in the standard library: a lot of functions have almost the same implementation between the `uniform`/`varying` versions and also sometimes between the `float16`, `float` and `double` input types with minor differences.

If I do not use templates (I only found ugly hacks to so far so to be able to mostly use them), then I expect the code to be possibly bigger than libdivide due to the higher number of combination coming from the variability (i.e. `uniform`/`varying`). This is very bug prone and hard to maintain due to the massive amount of repeated code with minor differences.

In C++, type traits and more specifically the following can help a lot to fix such issues (though C++ does not support `varying` variables): `std::conditional`, `std::is_signed`, `std::is_same`. It would be really great to be able to do such a thing in ISPC or something similar (and have something like `is_varying`/`is_uniform` so to build `uniform`/`varying` types based on this).

Using pre-processor's macros can certainly help to significantly reduce the size of the code (though the generated code will certainly be quite big), but I think this is not going to be great to read in the end...

(the post is bigger than expected since there is a lot more to say than I first thought).

**Comment by nurmukhametov** [comment]
> Yet another solution is to re-implement libdivide from scratch in ISPC. This solve most of the issue. However, this is not easy to do, it will certainly be a bit less efficient than the libdivide code using manually intrinsics.

Yes, I think this is the most reasonable one, as it gives us ultimate control over the implementation. Moreover, we can use intrinsics in ISPC (especially within builtins which we do frequently). In general, this should not result in any performance slowdown (otherwise it is a bug).

> One main issue is that there are many variants to generates:
> - uniform/varying numerator
> - uniform/varying denominator

Why do we need to consider these variants? I would expect that we only need to support `varying`/`varying` because 
if either operand is uniform, ISPC promotes it to varying  (see [link](https://godbolt.org/z/d6Excoe1G)).

> In C++, type traits and more specifically the following can help a lot to fix such issues (though C++ does not support varying variables): std::conditional, std::is_signed, std::is_same. It would be really great to be able to do such a thing in ISPC or something similar (and have something like is_varying/is_uniform so to build uniform/varying types based on this).

Yes, ISPC lacks these features. I think the separate issue is worth creating about this with your practical example (ideally small and self-contained).

**Comment by zephyr111** [comment]
> Yes, I think this is the most reasonable one, as it gives us ultimate control over the implementation. Moreover, we can use intrinsics in ISPC (especially within builtins which we do frequently). In general, this should not result in any performance slowdown (otherwise it is a bug).

I agree. That being said, for now, I do not plan to provide equal performance to libdivide but just something close in loops. Indeed, libidivide uses 128-bit operations or target-dependent optimization and doing that in ISPC is not straightforward, starting from the support of 128-bit integers which certainly far from being simple (or just a subset of 128-bit wide operations). 

> Why do we need to consider these variants? I would expect that we only need to support varying/varying because
if either operand is uniform, ISPC promotes it to varying (see [link](https://godbolt.org/z/d6Excoe1G)).

There are several reasons (mainly related to the divisor-related pre-computed structure).
- First, some functions (pre-computation) certainly will be faster for uniform divisor due to conditionals in it and certainly slower in SIMD instructions than their scalar alternatives (e.g. 64-bit multiplications resulting in 128-bit values), especially with target using double-pumping (or CPUs doing hardware double-pumping SIMD like IceLake/Zen4).
- Secondly, the divisor structure is written so if it is a varying one (while only one lane is actually useful) N times more memory is required for that. This is not much a problem as long as this is stored in registers, but it is a problem when such pre-computed state is stored in memory. Users can store only the first lane but this is a bit sad to request them to do that in this case. 
- Thirdly, I use references to mutate/read the structure (AFAIR a bit like ISPC random-related functions) and promotion from uniform to varying does not work in this case. I want to avoid copies if possible (though LLVM should optimize them if the code is inlined). 
- Besides, an efficient specialization could be done in the uniform/uniform case later.

By the way, just to clarify, my first goal is to provide functions to do optimized division in ISPC like libdivide does and not to to that automatically (though it can be done later). I am currently not convinced at all this is a good idea to do that automatically by default, especially since divisor 0 and 1 will not be supported (and supporting them correctly means to add more checks making the resulting code slower while this is fine to provide a custom function in the doc not supporting this as long as this is clearly documented and the function contains assertions for debugging purposes). Hence, promotion to `varying` in the current division binary operator have no impact on this (but does impact the case where we automatically replace the current divisions).

**Comment by nurmukhametov** [comment]
> By the way, just to clarify, my first goal is to provide functions to do optimized division in ISPC like libdivide does and not to to that automatically (though it can be done later).

Ah, right! I hadn't realized that your intention was to create separate functions for optimized integer divisions.

**Comment by MkazemAkhgary** [comment]
Wow, thanks for the detailed follow up! Personally, as an ISPC user, I’m fine with calling functions to handle fast division - even if it means manually passing the magic number around - instead of having ISPC do it automatically (though that would certainly be nice!).

Division by zero is undefined behavior in C and C++, so I think it shouldn't matter what happens if the divisor is zero. as for the case when it’s one, perhaps it could even be hinted with `assume(divisor != 1);` by the user to eliminate the branch.

**Comment by Andarwinux** [comment]
FYI, LLVM24's x86 backend implemented FP division based vector integer division lowering for i8/i16/i32/i64
